### PR TITLE
Add validation and operations tests for clearBuffer

### PR DESCRIPTION
**Pull request by toji**
<hr>

**Author checklist for test code/plans:**

- [x] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [x] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [x] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [ ] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [ ] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [ ] Existing (or new) test helpers are used where they would reduce complexity.
- [ ] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

File: src/webgpu/api/operation/command_buffer/clearBuffer.spec.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  - covers the end of the buffer
+  - covers neither the beginning nor the end of the buffer`
+  )
+  .paramsSubcasesOnly(u =>
+    u //
+      .combine('offset', [0, 4, 8, 16, undefined])
+      .combine('size', [0, 4, 8, 16, undefined])
+      .expand('bufferSize', p => [
+        (p.offset || 0) + (p.size || 16),
+        (p.offset || 0) + (p.size || 16) + 8,
```
Comment (kainino0x) [inline src/webgpu/api/operation/command_buffer/clearBuffer.spec.ts:26]: nit: Prefer the `??` ("nullish coalescing") operator over `||` for this

File: src/webgpu/api/operation/command_buffer/clearBuffer.spec.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      bufferData,
+      GPUBufferUsage.COPY_DST | GPUBufferUsage.COPY_SRC
+    );
+
+    const encoder = t.device.createCommandEncoder();
+    encoder.clearBuffer(buffer, offset, size);
+    t.device.queue.submit([encoder.finish()]);
+
+    const expectOffset = offset === undefined ? 0 : offset;
+    const expectSize = size === undefined ? bufferSize - expectOffset : size;
```
Comment (kainino0x) [inline src/webgpu/api/operation/command_buffer/clearBuffer.spec.ts:47]: `??` should work here too

File: src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    t.TestClearBuffer({
+      buffer: errorBuffer,
+      offset: 0,
+      size: 8,
+      isSuccess: false,
+    });
+  });
+
+g.test('default_args')
+  .desc(`Test that calling clearBuffer with a default offset and size is valid.`)
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:44]: Should also test with offset=undefined,size=value; and offset=value,size=undefined

File: src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    });
+  });
+
+g.test('size_alignment')
+  .desc(
+    `
+    Test that the clear size must be 4 byte aligned.
+    - Test size is not a multiple of 4.
+    - Test size is 0.
+    - Test size overflows the buffer size.
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:87]: I don't see a subcase for this. Looks like it's probably tested in `'overflow'` anyway.

File: src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+      isSuccess,
+    });
+  });
+
+g.test('offset_alignment')
+  .desc(
+    `
+    Test that the clear offsets must be 4 byte aligned.
+    - Test offset is not a multiple of 4.
+    - Test offset is larger than the buffer size.
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:120]: same

File: src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    t.TestClearBuffer({
+      buffer,
+      offset,
+      size,
+      isSuccess: false,
+    });
+  });
+
+g.test('out_of_bounds')
+  .desc(`Test that clears which exceed the buffer bounds invalid.`)
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:173]: nit:
```suggestion
  .desc(`Test that clears which exceed the buffer bounds are invalid.`)
```

File: src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+  .paramsSubcasesOnly([
+    { offset: 0, size: 32, _isSuccess: true },
+    { offset: 0, size: 36 },
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:176]: Should probably test edge cases like `{ offset: 32, size: 0 }, { offset: 32, size: 4 }`

**Review by kainino0x — APPROVED** [review]
LGTM with nits

File: package.json [hunk 8]
```diff
@@ -40,7 +40,7 @@
     "@types/offscreencanvas": "^2019.6.2",
     "@types/serve-index": "^1.9.1",
     "@typescript-eslint/parser": "^4.22.0",
-    "@webgpu/types": "^0.1.7",
+    "@webgpu/types": "^0.1.8",
```
Comment (kainino0x) [inline package.json:43]: Sorry, I caused a merge conflict in #848. Will need to change this to `0.1.8` (without the caret `^`).

Comment (toji) [reply package.json:43]: Actually, I've found that there's an error in 0.1.8. Fix incoming. Also, there's a few changes that need to be made to the CTS to roll to the new types so I think I'll handle that as a separate PR.

**Comment by toji** [comment]
Addressed feedback, but this PR is dependent of https://github.com/gpuweb/cts/pull/849, which is dependent on https://github.com/gpuweb/types/pull/87, so it's not going to clear the checks yet.

**Pull request merged**